Create a registry docker for developer testing

Since the registry is a central component of PDS system, many developers use it as an interface for their own development.

To spare them some time, we want to provide them with a docker image providing the elasticsearch data filled with some test data.

The docker image will:
- instantiate the elasticsearch server
- install and run harvest and registry-manager on test data as described on https://nasa-pds.github.io/pds-registry-app/install/index.html

The github repositories which are at stake are:
- https://github.com/NASA-PDS/pds-registry-app (the application package)
- https://github.com/NASA-PDS/harvest
- https://github.com/NASA-PDS/pds-registry-mgr-elastic

The docker resources will be manage in this pds-registry-app repository, in a specific directory (docker).

I am having problems with mvn linkcheck:linkcheck. Error is:

```
[INFO] Invoking the maven-site-plugin to ensure that all files are generated...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:07 min
[INFO] Finished at: 2021-02-04T19:13:01Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-linkcheck-plugin:1.2:linkcheck (default-cli) on project pds-registry-app: Execution default-cli of goal org.apache.maven.plugins:maven-linkcheck-plugin:1.2:linkcheck failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-linkcheck-plugin:1.2:linkcheck: java.lang.ExceptionInInitializerError: null
[ERROR] -----------------------------------------------------
```

I have googled it and it is not obvious on how to make it available like `mvn install linkcheck` but rather some changes to the POM. Therefore, I need to ask should I install it or is the README.md in this repo out of date?

--- a/Dockerfile.local
+++ b/Dockerfile.local
@@ -1,0 +1,49 @@
+FROM ubuntu:20.04
+
+# Get arguments from the build command line
+ARG version_reg_app
+ENV REGISTRY_MANAGER_HOME=/var/local/registry
+ENV VERSION_REG_APP=$version_reg_app
+
+# Build up the OS
+RUN export DEBIAN_FRONTEND=noninteractive && \
+    apt-get update && \
+    apt-get install -y curl \
+                       maven \
+                       openjdk-11-jdk-headless \
+                       tar
+
+# Make room for the app
+RUN mkdir /usr/local/registry-app-${VERSION_REG_APP}
+
+# Copy the data into the building container
+COPY LICENSE.txt /usr/local/registry-app-${VERSION_REG_APP}
+COPY NOTICE.txt /usr/local/registry-app-${VERSION_REG_APP}
+COPY README-PACKAGE.txt /usr/local/registry-app-${VERSION_REG_APP}
+COPY pom.xml /usr/local/registry-app-${VERSION_REG_APP}
+COPY settings.xml /usr/local/registry-app-${VERSION_REG_APP}
+COPY src /usr/local/registry-app-${VERSION_REG_APP}/src
+
+# Build the application and deploy it inside the container
+RUN cd /usr/local/registry-app-${VERSION_REG_APP} && \
+    mvn compile && \
+    mvn site && \
+    mvn package
+
+# Build up all of the bin files for general use
+RUN mkdir -p /usr/local/dist \
+             /var/local/registry/elastic \
+             /var/local/harvest/{archive,output} && \
+    cd /usr/local/registry-app-${VERSION_REG_APP} && \
+    cp target/harvest-*/bin/harvest \
+       target/registry-manager-*/bin/registry-manager \
+       /usr/local/bin/ && \
+    cp target/harvest-*/dist/* \
+       target/registry-manager-*/dist/* \
+       /usr/local/dist/ && \
+    cp target/registry-manager-4.0.2/elastic/* /var/local/registry/elastic/ && \
+    ln -s /var/local/registry/elastic /usr/local && \
+    cp -r target/harvest-*/conf /var/local/harvest/ && \
+    ( sed -e 's/\/path\/to\/archive/\/var\/local\/harvest\/archive/' \
+          target/harvest-*/conf/examples/registry.xml > \
+          /var/local/harvest/conf/examples/registry.xml )

--- a/Dockerfile.local
+++ b/Dockerfile.local
@@ -33,7 +33,8 @@ RUN cd /usr/local/registry-app-${VERSION_REG_APP} && \
 # Build up all of the bin files for general use
 RUN mkdir -p /usr/local/dist \
              /var/local/registry/elastic \
-             /var/local/harvest/{archive,output} && \
+             /var/local/harvest/archive \
+             /var/local/harvest/output && \
     cd /usr/local/registry-app-${VERSION_REG_APP} && \
     cp target/harvest-*/bin/harvest \
        target/registry-manager-*/bin/registry-manager \
@@ -46,4 +47,5 @@ RUN mkdir -p /usr/local/dist \
     cp -r target/harvest-*/conf /var/local/harvest/ && \
     ( sed -e 's/\/path\/to\/archive/\/var\/local\/harvest\/archive/' \
           target/harvest-*/conf/examples/registry.xml > \
-          /var/local/harvest/conf/examples/registry.xml )
+          /var/local/harvest/conf/examples/registry.xml ) && \
+    cp -r target/v11400/dph_example_archive/* /var/local/harvest/archive

--- a/README.md
+++ b/README.md
@@ -50,8 +50,10 @@ docker build --build-arg version_reg_app=$(git rev-parse HEAD) \
 
 ### Elastic Search
 
-1. prepare `mkdir /tmp/es`
-1. fetch `docker pull elasticsearch:7.10.1`
+1. prepare `mkdir /tmp/es /tmp/output`  
+    These two directories are to store the data between steps. If there is an existng elastic search database or harvest working area, then there is no need for this step.
+1. fetch `docker pull elasticsearch:7.10.1`  
+    Only need to fetch the image once.
 1. run  
     ```
     docker run --detach \
@@ -61,19 +63,79 @@ docker build --build-arg version_reg_app=$(git rev-parse HEAD) \
                --publish 9200:9200 \
                --publish 9300:9300 \
                --rm \
+               --user $UID \
                --volume /tmp/es:/usr/share/elasticsearch/data \
                elasticsearch:7.10.1
-    ```
+    ```  
+    Start the elastic search engine and make sure that any of the data created is owned by yourself so that you can clean it up. If you are not using the /tmp/es directory created above, then make sure to change the --volume argument as needed so that elastic search has access to it. Also change the --user arguement such that elastic search has full privledges to read, write, and modify the data.
 
-### Install test data
+### Install test data 
+
+This section will use the small test data set contained within the docker image. The steps used to do this can be used as a template to install your own more complicated datasets as well.
+
+First, install the registry schema for elastic search. Things to note, the URL host name is the --name argument value used when starting elastic search.
 
 ```
-docker run --network pds --rm pds_registry_app:$(git rev-parse HEAD) \
-    registry-manager create-registry -es http://es:9200 \
-                                     -schema /var/local/registry/elastic/registry.json 
+docker run --network pds \
+           --rm \
+           --user $UID \
+           pds_registry_app:$(git rev-parse HEAD) \
+              registry-manager create-registry -es http://es:9200 \
+                                               -schema /var/local/registry/elastic/registry.json 
 ```
 
+Second we need to harvest the test data. It is currently setup in /var/local/harvest/archive. If one wanted to use their own archive of data, they would simply have to add `--volume /path/to/archive:/var/local/harvest/archive` to the docker run portion of this command. Because we cannot easily run two commands in one docker run invokation, it is necessary to use an external directory for /var/local/havest/output. The ephemeral nature of a container erases what we would write into /var/local/harvest/output between the two docker run invokations.
 ```
-docker run --network pds --rm pds_registry_app:$(git rev-parse HEAD) \         harvest -c /var/local/harvest/conf/examples/registry.xml \
-                 -o /var/local/harvest/output
+docker run --network pds \
+           --rm \
+           --user $UID \
+           --volume /tmp/output:/var/local/harvest/output \
+           pds_registry_app:$(git rev-parse HEAD) harvest \
+              -c /var/local/harvest/conf/examples/registry.xml \
+              -o /var/local/harvest/output
 ```
+
+Third, and last, we need to now ingest the havested data from above into elastic search. Things to note: One, the URL host name is the --name argument value used when starting elastic search. Two, the output volume must be the same mapping as in the previous step.
+```
+docker run --network pds \
+           --rm \
+           --user $UID \
+           --volume /tmp/output:/var/local/harvest/output \
+           pds_registry_app:$(git rev-parse HEAD) \
+           registry-manager load-data \
+              -es http://es:9200 \
+              -file /var/local/harvest/output
+```
+
+### Verify it worked
+
+The verification script is written in Python 3.x with requests installed as well only because that is what the person doing the work knows the best. A simple curl then lots of scrolling and eyeballing can do the same thing.
+
+
+```
+python3 <<EOF
+import requests
+response = requests.get (' http://localhost:9200/registry/_search?q=*')
+if response.status_code == 200:
+    print ('successfully found elastic search')
+    data = response.json()
+
+    if not data['timed_out']:
+        print ('successfully searched for query')
+
+        if data['hits']['total']['value'] == 17:
+            print ('successfully found all expected data')
+        else: print ('FAILED to find all expected data')
+    else: print ('FAILED query search')
+else: print ('FAILED to find elastic search')
+EOF
+```
+
+Output should look like this if everything is working:
+```
+successfully found elastic search
+successfully searched for query
+successfully found all expected data
+
+```
+

--- a/README.md
+++ b/README.md
@@ -33,3 +33,47 @@ Prepare the package:
     mvn github-release:github-release  ## publish the package on github  # TO BE DONE REPLACE WITH CONMAN or GITHUB ACTION
 
     
+# Docker
+
+## Build
+
+### Local git version
+
+```
+docker build --build-arg version_reg_app=$(git rev-parse HEAD) \
+             --file Dockerfile.local \
+             --tag pds_registry_app:$(git rev-parse HEAD) \
+             .
+```
+
+## Run
+
+### Elastic Search
+
+1. prepare `mkdir /tmp/es`
+1. fetch `docker pull elasticsearch:7.10.1`
+1. run  
+    ```
+    docker run --detach \
+               --env "discovery.type=single-node" \
+               --name es \
+               --network pds \
+               --publish 9200:9200 \
+               --publish 9300:9300 \
+               --rm \
+               --volume /tmp/es:/usr/share/elasticsearch/data \
+               elasticsearch:7.10.1
+    ```
+
+### Install test data
+
+```
+docker run --network pds --rm pds_registry_app:$(git rev-parse HEAD) \
+    registry-manager create-registry -es http://es:9200 \
+                                     -schema /var/local/registry/elastic/registry.json 
+```
+
+```
+docker run --network pds --rm pds_registry_app:$(git rev-parse HEAD) \         harvest -c /var/local/harvest/conf/examples/registry.xml \
+                 -o /var/local/harvest/output
+```

